cosmos/precompile/erc20/middleware.go

```diff
@@ -0,0 +1,166 @@
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package erc20
+
+import (
+	"context"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	cosmlib "pkg.berachain.dev/polaris/cosmos/lib"
+	erc20types "pkg.berachain.dev/polaris/cosmos/x/erc20/types"
+	"pkg.berachain.dev/polaris/eth/common"
+	ethprecompile "pkg.berachain.dev/polaris/eth/core/precompile"
+)
+
+// convertCoinToERC20 converts SDK coins to ERC20 tokens for an owner.
+func (c *Contract) convertCoinToERC20(
+	ctx context.Context,
+	caller common.Address,
+	evm ethprecompile.EVM,
+	value *big.Int,
+	denom string,
+	owner common.Address,
+	amount *big.Int,
+) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	// burn amount SDK coins from owner
+	err := cosmlib.BurnCoinsFromAddress(
+		sdkCtx, c.bk, erc20types.ModuleName, owner, denom, amount,
+	)
+	if err != nil {
+		return err
+	}
+
+	// get ERC20 token address pairing with SDK coin denomination
+	resp, err := c.em.ERC20AddressForCoinDenom(
+		ctx, &erc20types.ERC20AddressForCoinDenomRequest{
+			Denom: denom,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	var token common.Address
+	if resp.Token == "" {
+		// deploy the new ERC20 token contract (deployer of this contract is the ERC20 module!)
+		token, err = c.deployNewERC20Contract(sdkCtx, evm, c.RegistryKey(), denom, value)
+		if err != nil {
+			return err
+		}
+
+		// create the new ERC20 token contract pairing with SDK coin denomination
+		c.em.RegisterCoinERC20Pair(sdkCtx, denom, token)
+
+		// mint amount ERC20 tokens to the owner
+		if err = c.callERC20mint(sdkCtx, evm, c.RegistryKey(), token, owner, amount); err != nil {
+			return err
+		}
+
+		sdkCtx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				erc20types.EventTypeConvertCoinToERC20,
+				sdk.NewAttribute(erc20types.AttributeKeyDenom, denom),
+				sdk.NewAttribute(erc20types.AttributeKeyToken, token.Hex()),
+			),
+		)
+		return nil
+	}
+
+	// convert ERC20 token bech32 address to common.Address
+	tokenAcc, err := sdk.AccAddressFromBech32(resp.Token)
+	if err != nil {
+		return err
+	}
+	token = cosmlib.AccAddressToEthAddress(tokenAcc)
+
+	// transfer amount ERC20 tokens from ERC20 module precompile contract to owner
+	if err = c.callERC20transferFrom(sdkCtx, evm, caller, token, c.RegistryKey(), owner, amount); err != nil {
+		return err
+	}
```

Comment:
transfer is correct here? the module precompile contract should already have received these tokens at some point.